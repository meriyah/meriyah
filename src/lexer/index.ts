export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
