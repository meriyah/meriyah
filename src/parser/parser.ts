import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
