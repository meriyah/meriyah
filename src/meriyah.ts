import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
