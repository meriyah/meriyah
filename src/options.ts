import type * as ESTree from './estree';
import { pushComment, pushToken } from './parser/parser';
import { type Token } from './token';

/**
 * Function calls when semicolon inserted.
 */
type OnInsertedSemicolon = (pos: number) => any;

/**
 * Token process function.
 */
export type OnToken = (token: string, start: number, end: number, loc: ESTree.SourceLocation) => any;

/**
 * Comment process function.
 */
export type OnComment = (
  type: ESTree.CommentType,
  value: string,
  start: number,
  end: number,
  loc: ESTree.SourceLocation,
) => any;

/**
 * The parser options.
 */
export interface Options {
  // Allow module code
  module?: boolean;
  // Enable stage 3 support (ESNext)
  next?: boolean;
  // Enable start and end offsets to each node
  ranges?: boolean;
  // Enable web compatibility
  webcompat?: boolean;
  // Enable line/column location information to each node
  loc?: boolean;
  // Attach raw property to each literal and identifier node
  raw?: boolean;
  // Allow return in the global scope
  globalReturn?: boolean;
  // Enable implied strict mode
  impliedStrict?: boolean;
  // Enable non-standard parenthesized expression node
  preserveParens?: boolean;
  // Enable lexical binding and scope tracking
  lexical?: boolean;
  // Adds a source attribute in every nodeâ€™s loc object when the locations option is `true`
  source?: string;
  // Enable React JSX parsing
  jsx?: boolean;
  // Allows comment extraction. Accepts either a callback function or an array
  onComment?: ESTree.Comment[] | OnComment;
  // Allows detection of automatic semicolon insertion. Accepts a callback function that will be passed the character offset where the semicolon was inserted
  onInsertedSemicolon?: OnInsertedSemicolon;
  // Allows token extraction. Accepts either a callback function or an array
  onToken?: Token[] | OnToken;
}

export type NormalizedOptions = Omit<Options, 'onComment' | 'onToken'> & {
  onComment?: OnComment;
  onToken?: OnToken;
};

export function normalizeOptions(rawOptions: Options): NormalizedOptions {
  const options = { ...rawOptions } as NormalizedOptions;

  // Accepts either a callback function to be invoked or an array to collect comments (as the node is constructed)
  if (options.onComment) {
    options.onComment = Array.isArray(options.onComment) ? pushComment(options.onComment, options) : options.onComment;
  }

  // Accepts either a callback function to be invoked or an array to collect tokens
  if (options.onToken) {
    options.onToken = Array.isArray(options.onToken) ? pushToken(options.onToken, options) : options.onToken;
  }

  return options;
}
