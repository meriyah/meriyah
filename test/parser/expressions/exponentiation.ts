import * as t from 'node:assert/strict';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail, pass } from '../../test-utils';

describe('Expressions - Exponentiation', () => {
  for (const arg of [
    'delete O.p ** 10',
    'delete x ** 10',
    '~O.p ** 10',
    '~x ** 10',
    '!O.p ** 10',
    '!x ** 10',
    '+O.p ** 10',
    '+x ** 10',
    '-O.p ** 10',
    '-x ** 10',
    '!1 ** 2',
    'void 1 ** 2;',
    'typeof 1 ** 2;',
    'typeof O.p ** 10',
    'typeof x ** 10',
    'void ** 10',
    'void O.p ** 10',
    'void x ** 10',
    '-x ** y',
    '++delete O.p ** 10',
    '--delete O.p ** 10',
    '++~O.p ** 10',
    '++~x ** 10',
    '--!O.p ** 10',
    '--!x ** 10',
    '++-O.p ** 10',
    '++-x ** 10',
    '--+O.p ** 10',
    '--+x ** 10',
    '[ x ] **= [ 2 ]',
    '[ x **= 2 ] = [ 2 ]',
    '{ x } **= { x: 2 }',
    '{ x: x **= 2 ] = { x: 2 }',
  ]) {
    it(`let O = { p: 1 }, x = 10; ; if (${arg}) { foo(); }`, () => {
      t.throws(() => {
        parseSource(`let O = { p: 1 }, x = 10; ; if (${arg}) { foo(); }`, { next: true, sourceType: 'module' });
      });
    });

    it(`var O = { p: 1 }, x = 10; ; (${arg})`, () => {
      t.throws(() => {
        parseSource(`var O = { p: 1 }, x = 10; ; (${arg})`, { next: true, impliedStrict: true });
      });
    });

    it(`var O = { p: 1 }, x = 10; foo(${arg})`, () => {
      t.throws(() => {
        parseSource(`var O = { p: 1 }, x = 10; foo(${arg})`, { webcompat: true });
      });
    });
  }

  fail('Expressions - Exponentiation (fail)', [
    '+1 ** 2;',
    '-3 ** 2;',
    '!1 ** 2;',
    'delete o.p ** 2;',
    '~3 ** 2;',
    '(a * +a ** a ** 3)',
    'typeof 3 ** 2;',
    'delete 3 ** 2;',
    '!3 ** 2;',
    '-x ** 2;',
    '+x ** 2;',
    '(~3 ** 2)',
    '(typeof 3 ** 2)',
    '(delete 3 ** 2)',
    '(!3 ** 2)',
    '(+x ** 2)',
    '(a * +a ** a ** 3)',
    'for (var import.meta of [1]) {}',
    'async function f() { await 2 ** 2; }',
  ]);

  for (const arg of [
    '(delete O.p) ** 10',
    '(~O.p) ** 10',
    '(~x) ** 10',
    '(!O.p) ** 10',
    '(!x) ** 10',
    '(+O.p) ** 10',
    '(+x) ** 10',
    '(-O.p) ** 10',
    'x ** y ** z',
    '++x ** y',
    '(-x) ** y',
    '-(x ** y)',
    '(-x) ** 10',
    '(typeof O.p) ** 10',
    '(typeof x) ** 10',
    '(void 0) ** 10',
    '(void O.p) ** 10',
    '(void x) ** 10',
    '2 ** ++exponent, 8',
    '2 ** -1 * 2, 1',
    '2 ** (3 ** 2)',
    '2 ** 3 ** 2, 512',
    '16 / 2 ** 2, 4',
    '++O.p ** 10',
    '++x ** 10',
    '--O.p ** 10',
    '--base ** 2',
    '2 ** !s',
    '2 ** +n',
    '!(3 ** 2)',
    '-(3 ** 2)',
    '--x ** 10',
    'O.p++ ** 10',
    'x++ ** 10',
    'O.p-- ** 10',
    'x-- ** 10',
  ]) {
    it(`var O = { p: 1 }, x = 10; ; if (${arg}) { foo(); }`, () => {
      t.doesNotThrow(() => {
        parseSource(`var O = { p: 1 }, x = 10; ; if (${arg}) { foo(); }`, { next: true, sourceType: 'module' });
      });
    });

    it(`var O = { p: 1 }, x = 10; ; (${arg})`, () => {
      t.doesNotThrow(() => {
        parseSource(`var O = { p: 1 }, x = 10; ; (${arg})`, { next: true, impliedStrict: true });
      });
    });

    it(`var O = { p: 1 }, x = 10; foo(${arg})`, () => {
      t.doesNotThrow(() => {
        parseSource(`var O = { p: 1 }, x = 10; foo(${arg})`, { webcompat: true });
      });
    });
  }

  pass('Expressions - Exponentiation (pass)', [
    { code: '(base **= 3) === -27', options: { ranges: true } },
    { code: '2 ** 4', options: { ranges: true } },
    { code: 'new x ** 2;', options: { ranges: true } },
    'true ** a',
    '++x ** a',
    '--x ** a',
    'x++ ** a',
    'x-- ** a',
    '+a * b ** c ** 3',
    '(2 ** 4)',
    { code: '(new x ** 2)', options: { ranges: true } },
    '(true ** a)',
    { code: '(++x ** a)', options: { ranges: true } },
    '(+c * b ** a ** 3)',
    '(+1) ** 2',
    'async function f() { (await 2) ** 2; }',
    'async function f() { await (2 ** 2); }',
  ]);
});
