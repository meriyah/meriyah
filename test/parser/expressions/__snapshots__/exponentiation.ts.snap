// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > !1 ** 2; 1`] = `
"SyntaxError [1:3-1:5]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | !1 ** 2;
    |    ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > !3 ** 2; 1`] = `
"SyntaxError [1:3-1:5]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | !3 ** 2;
    |    ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > (!3 ** 2) 1`] = `
"SyntaxError [1:4-1:6]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | (!3 ** 2)
    |     ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > (+x ** 2) 1`] = `
"SyntaxError [1:4-1:6]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | (+x ** 2)
    |     ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > (~3 ** 2) 1`] = `
"SyntaxError [1:4-1:6]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | (~3 ** 2)
    |     ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > (a * +a ** a ** 3) 1`] = `
"SyntaxError [1:8-1:10]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | (a * +a ** a ** 3)
    |         ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > (a * +a ** a ** 3) 2`] = `
"SyntaxError [1:8-1:10]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | (a * +a ** a ** 3)
    |         ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > (delete 3 ** 2) 1`] = `
"SyntaxError [1:10-1:12]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | (delete 3 ** 2)
    |           ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > (typeof 3 ** 2) 1`] = `
"SyntaxError [1:10-1:12]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | (typeof 3 ** 2)
    |           ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > +1 ** 2; 1`] = `
"SyntaxError [1:3-1:5]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | +1 ** 2;
    |    ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > +x ** 2; 1`] = `
"SyntaxError [1:3-1:5]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | +x ** 2;
    |    ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > ~3 ** 2; 1`] = `
"SyntaxError [1:3-1:5]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | ~3 ** 2;
    |    ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > -3 ** 2; 1`] = `
"SyntaxError [1:3-1:5]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | -3 ** 2;
    |    ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > -x ** 2; 1`] = `
"SyntaxError [1:3-1:5]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | -x ** 2;
    |    ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > async function f() { await 2 ** 2; } 1`] = `
"SyntaxError [1:29-1:31]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | async function f() { await 2 ** 2; }
    |                              ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > delete 3 ** 2; 1`] = `
"SyntaxError [1:9-1:11]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | delete 3 ** 2;
    |          ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > delete o.p ** 2; 1`] = `
"SyntaxError [1:11-1:13]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | delete o.p ** 2;
    |            ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > for (var import.meta of [1]) {} 1`] = `
"SyntaxError [1:9-1:15]: Invalid keyword
> 1 | for (var import.meta of [1]) {}
    |          ^^^^^^ Invalid keyword"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (fail) > typeof 3 ** 2; 1`] = `
"SyntaxError [1:9-1:11]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses
> 1 | typeof 3 ** 2;
    |          ^^ Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses"
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > (++x ** a) 1`] = `
{
  "body": [
    {
      "end": 10,
      "expression": {
        "end": 9,
        "left": {
          "argument": {
            "end": 4,
            "name": "x",
            "range": [
              3,
              4,
            ],
            "start": 3,
            "type": "Identifier",
          },
          "end": 4,
          "operator": "++",
          "prefix": true,
          "range": [
            1,
            4,
          ],
          "start": 1,
          "type": "UpdateExpression",
        },
        "operator": "**",
        "range": [
          1,
          9,
        ],
        "right": {
          "end": 9,
          "name": "a",
          "range": [
            8,
            9,
          ],
          "start": 8,
          "type": "Identifier",
        },
        "start": 1,
        "type": "BinaryExpression",
      },
      "range": [
        0,
        10,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 10,
  "range": [
    0,
    10,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > (+1) ** 2 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "argument": {
            "type": "Literal",
            "value": 1,
          },
          "operator": "+",
          "prefix": true,
          "type": "UnaryExpression",
        },
        "operator": "**",
        "right": {
          "type": "Literal",
          "value": 2,
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > (+c * b ** a ** 3) 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "argument": {
            "name": "c",
            "type": "Identifier",
          },
          "operator": "+",
          "prefix": true,
          "type": "UnaryExpression",
        },
        "operator": "*",
        "right": {
          "left": {
            "name": "b",
            "type": "Identifier",
          },
          "operator": "**",
          "right": {
            "left": {
              "name": "a",
              "type": "Identifier",
            },
            "operator": "**",
            "right": {
              "type": "Literal",
              "value": 3,
            },
            "type": "BinaryExpression",
          },
          "type": "BinaryExpression",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > (2 ** 4) 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "type": "Literal",
          "value": 2,
        },
        "operator": "**",
        "right": {
          "type": "Literal",
          "value": 4,
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > (base **= 3) === -27 1`] = `
{
  "body": [
    {
      "end": 20,
      "expression": {
        "end": 20,
        "left": {
          "end": 11,
          "left": {
            "end": 5,
            "name": "base",
            "range": [
              1,
              5,
            ],
            "start": 1,
            "type": "Identifier",
          },
          "operator": "**=",
          "range": [
            1,
            11,
          ],
          "right": {
            "end": 11,
            "range": [
              10,
              11,
            ],
            "start": 10,
            "type": "Literal",
            "value": 3,
          },
          "start": 1,
          "type": "AssignmentExpression",
        },
        "operator": "===",
        "range": [
          0,
          20,
        ],
        "right": {
          "argument": {
            "end": 20,
            "range": [
              18,
              20,
            ],
            "start": 18,
            "type": "Literal",
            "value": 27,
          },
          "end": 20,
          "operator": "-",
          "prefix": true,
          "range": [
            17,
            20,
          ],
          "start": 17,
          "type": "UnaryExpression",
        },
        "start": 0,
        "type": "BinaryExpression",
      },
      "range": [
        0,
        20,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 20,
  "range": [
    0,
    20,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > (new x ** 2) 1`] = `
{
  "body": [
    {
      "end": 12,
      "expression": {
        "end": 11,
        "left": {
          "arguments": [],
          "callee": {
            "end": 6,
            "name": "x",
            "range": [
              5,
              6,
            ],
            "start": 5,
            "type": "Identifier",
          },
          "end": 6,
          "range": [
            1,
            6,
          ],
          "start": 1,
          "type": "NewExpression",
        },
        "operator": "**",
        "range": [
          1,
          11,
        ],
        "right": {
          "end": 11,
          "range": [
            10,
            11,
          ],
          "start": 10,
          "type": "Literal",
          "value": 2,
        },
        "start": 1,
        "type": "BinaryExpression",
      },
      "range": [
        0,
        12,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 12,
  "range": [
    0,
    12,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > (true ** a) 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "type": "Literal",
          "value": true,
        },
        "operator": "**",
        "right": {
          "name": "a",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > ++x ** a 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "argument": {
            "name": "x",
            "type": "Identifier",
          },
          "operator": "++",
          "prefix": true,
          "type": "UpdateExpression",
        },
        "operator": "**",
        "right": {
          "name": "a",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > +a * b ** c ** 3 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "argument": {
            "name": "a",
            "type": "Identifier",
          },
          "operator": "+",
          "prefix": true,
          "type": "UnaryExpression",
        },
        "operator": "*",
        "right": {
          "left": {
            "name": "b",
            "type": "Identifier",
          },
          "operator": "**",
          "right": {
            "left": {
              "name": "c",
              "type": "Identifier",
            },
            "operator": "**",
            "right": {
              "type": "Literal",
              "value": 3,
            },
            "type": "BinaryExpression",
          },
          "type": "BinaryExpression",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > --x ** a 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "argument": {
            "name": "x",
            "type": "Identifier",
          },
          "operator": "--",
          "prefix": true,
          "type": "UpdateExpression",
        },
        "operator": "**",
        "right": {
          "name": "a",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > 2 ** 4 1`] = `
{
  "body": [
    {
      "end": 6,
      "expression": {
        "end": 6,
        "left": {
          "end": 1,
          "range": [
            0,
            1,
          ],
          "start": 0,
          "type": "Literal",
          "value": 2,
        },
        "operator": "**",
        "range": [
          0,
          6,
        ],
        "right": {
          "end": 6,
          "range": [
            5,
            6,
          ],
          "start": 5,
          "type": "Literal",
          "value": 4,
        },
        "start": 0,
        "type": "BinaryExpression",
      },
      "range": [
        0,
        6,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 6,
  "range": [
    0,
    6,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > async function f() { (await 2) ** 2; } 1`] = `
{
  "body": [
    {
      "async": true,
      "body": {
        "body": [
          {
            "expression": {
              "left": {
                "argument": {
                  "type": "Literal",
                  "value": 2,
                },
                "type": "AwaitExpression",
              },
              "operator": "**",
              "right": {
                "type": "Literal",
                "value": 2,
              },
              "type": "BinaryExpression",
            },
            "type": "ExpressionStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "generator": false,
      "id": {
        "name": "f",
        "type": "Identifier",
      },
      "params": [],
      "type": "FunctionDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > async function f() { await (2 ** 2); } 1`] = `
{
  "body": [
    {
      "async": true,
      "body": {
        "body": [
          {
            "expression": {
              "argument": {
                "left": {
                  "type": "Literal",
                  "value": 2,
                },
                "operator": "**",
                "right": {
                  "type": "Literal",
                  "value": 2,
                },
                "type": "BinaryExpression",
              },
              "type": "AwaitExpression",
            },
            "type": "ExpressionStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "generator": false,
      "id": {
        "name": "f",
        "type": "Identifier",
      },
      "params": [],
      "type": "FunctionDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > new x ** 2; 1`] = `
{
  "body": [
    {
      "end": 11,
      "expression": {
        "end": 10,
        "left": {
          "arguments": [],
          "callee": {
            "end": 5,
            "name": "x",
            "range": [
              4,
              5,
            ],
            "start": 4,
            "type": "Identifier",
          },
          "end": 5,
          "range": [
            0,
            5,
          ],
          "start": 0,
          "type": "NewExpression",
        },
        "operator": "**",
        "range": [
          0,
          10,
        ],
        "right": {
          "end": 10,
          "range": [
            9,
            10,
          ],
          "start": 9,
          "type": "Literal",
          "value": 2,
        },
        "start": 0,
        "type": "BinaryExpression",
      },
      "range": [
        0,
        11,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 11,
  "range": [
    0,
    11,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > true ** a 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "type": "Literal",
          "value": true,
        },
        "operator": "**",
        "right": {
          "name": "a",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > x++ ** a 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "argument": {
            "name": "x",
            "type": "Identifier",
          },
          "operator": "++",
          "prefix": false,
          "type": "UpdateExpression",
        },
        "operator": "**",
        "right": {
          "name": "a",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Exponentiation > Expressions - Exponentiation (pass) > x-- ** a 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "argument": {
            "name": "x",
            "type": "Identifier",
          },
          "operator": "--",
          "prefix": false,
          "type": "UpdateExpression",
        },
        "operator": "**",
        "right": {
          "name": "a",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;
